### ISSUE TITLE
Proposal: generalize extension to search in other websites

### ISSUE THREAD

**Issue by infokiller** [infokiller/web-search-navigator#82]
The idea is to support other search engines such as Bing, Amazon, Github, etc. 
I envision the code structured in a way that adding basic support for a new website only involves specifying things like the search results selectors, search box selector, etc.
Some of the current features are very specific to Google, and my intention is to keep supporting and enhancing them. This will also need to be reflected in the user options, which will include general navigation options (such as keybinding for next/previous search result and getting back to the search box), and options specific to a website.

@alexbarrett @mancze @rik-smeets 
I'm curious to hear your thoughts.

**Comment by rik-smeets** [comment]
I'm a proponent of making the extension more generic, since I don't only use Google but also other search engines from time to time. Also, keyboard navigation on sites like YouTube is pretty bad. Supporting that as well could be great! I think we could gradually restructure the code where necessary to support this. 

We should however beware that we should not try to fit everything in this extension. This could make it harder to maintain and understand. For some websites, for example, a separate extension might be the better option.

**Comment by mancze** [comment]
I don't mind that this extension is specific to google. I use only Google search engine so it's a fit for me. I wouldn't benefit from supporting other search engines. Also I think its name helps to discover the extension for a new users - as it did for me. Contrary I fear it would make things more complicated (more complicated settings) and gave feeling of "bloated" extension. That's my opinion as an user.

From developer's point of view I see no problem with supporting other search engines as well. You pointed pretty well how it should be implemented.

The pragmatist in me tells me that it implies more work to maintain. And I have very few spare time to invest in things like this. Since this extension is not super popular extra work might not pay itself off when it won't find it's user-base. And we did not receive much requests for adding support of other search engines (yes, I'm aware it's kinda the chicken or the egg problem).

And finally: The playful soul in me already thought about proposed concept. And I came to the idea of completely another extension not being even search engine specific nor search specific. I don't want to share more details for now as I need to think more deeply about it. But currently it's the actual way of my thinking.

TLDR: If you think it's a way to go I consent. Yet I do not plan to invest my time for support of another search engines.

And last idea for an end: Shipping several editions of the same extension might solve the specificity-generalization dilemma. And by that I mean that build process might produce version for each supported search engine separately and one version which would cover all. Just an idea. It means complications and feels a little over-engineered but I wanted to share that.

**Comment by infokiller** [comment]
Thanks for the feedback guys!
A few notes:

- It may indeed make sense to split the extension at the release level (but they can still share most of the code). I'm not worried about code bloat. What i'm more worried about is a bloat in the permissions for the extension. For example, a user that only uses google shouldn't need to give permissions for youtube or amazon, and that may create a privacy issue for some users (which I totally understand). I'll try to explore this further to get a better understanding of the tradeoffs involved.
- I'm not much worried about insufficient support for Firefox/Edge-  I believe that if we clearly state that the Firefox version doesn't officially support other search engines it should solve most of the support issues. I know that even if this functionality is turned off or marked as unsupported, it may still increase the likelihood of bugs in the firefox version, but I believe we can mitigate this with reasonable engineering.

**Comment by henriklb** [comment]
I'd love this addon to be duckduckgo compatible. While using duckduckgo I miss the 'v', 'i' etc. shortcuts

**Comment by LaPeste** [comment]
Hi, any news on the status of this?
I've been waiting so much for this extension to work with StartPage.

I'd be great if you could share with us your expectations about this.  :)

Cheers

**Comment by infokiller** [comment]
Hey, no news yet. Working on this, along with other improvements, is on my backlog, but I can't say if it will take a month, a year, or a decade...

**Comment by LaPeste** [comment]
Alright, thank you very much for letting us know.

Out of curiosity I built it from source and tried it out.That pesky anchor being undefined :P 

Jokes aside, thank you a lot for all the work.

Cheers

**Comment by infokiller** [comment]
The current master branch has optional support for YouTube and Startpage (needs to be enabled in the options page).
If I'll have time, I may add support for Google Scholar, Github, or something else.
Closing this now, though the versions released on the browser stores are not updated yet.

**Comment by LaPeste** [comment]
I built from source, loaded as unpacked and enabled the setting for startpage, but it doesn't work for me.
I use Brave 1.10.93 based on Chromium 83.0.4103.106 on Ubuntu.

Is there more that I should do?

**Comment by infokiller** [comment]
@LaPeste you need to enable it in the options page (right click extension icon, then options)
I think the error you see is OK (chrome not happy about firefox specific setting).
BTW the best way to test is using `web-ext` (doesn't affect your current browser profile and opens a new browser profile with only this extension installed), see https://github.com/infokiller/web-search-navigator#building-for-development

**Comment by LaPeste** [comment]
When I wrote "enabled the setting for startpage" I meant that I went in the options as recommended by you and enabled startpage. But it doesn't work. Do you think that using 'web-ext' would help with the current problem? How?

**Comment by infokiller** [comment]
I'm opening a new issue to track this.
